### PR TITLE
Campaign summary statistics (performance boost and date range specificity)

### PR DESCRIPTION
**Pull request by heathdutton**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | Yes
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | No
| Deprecations? | No

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

#### ToDo

- [ ] Write more functional tests to guarantee the charts and counts work as they should.
- [ ] Fix the command as it seems it does not count contacts properly. There is a functional test that shows it.
- [ ] Implement the Pending Count that has been added after this PR was submitted. It shows bad counts because of it after rebase.

[//]: # ( Required: )
#### Description:
**TL;DR Make big campaigns practical.**

If you have a campaign with 20 events, and 1700 leads a day, that's over a million new rows, per month and if the campaign is still active the queries on the campaign view page will take several minutes on a good day.

Even with various recent improvements, the problem is still growing exponentially as you scale up. This PR aims to fix the problem permanently by summarizing historical campaign data by hours (as services like Google Analytics do). The view will behave exactly the same (same charts, same tabular data, no delays), but will have drastically fewer rows to crunch through to generate the data (no joins, just one flat table with some sums). 

A command line function `mautic:campaign:summarize` is provided to perform an optional one-time backfill of the summary data. This might be needed by someone who wants to turn on campaign summary data *and* needs it to be retroactive. The command runs backward through time (current data is usually more pertinent), filling in missing summary data till it's done. It handles about 10.5 million campaign event log entries per hour in my testing. It's not super speedy with 200+mil. events, but it's a one-time thing. This command can also be used to true-up data in case you are experiencing exceptions within your campaign cron tasks.

I've merged in and refactored work from #6021 (Optionally display campaign data by date range) for a number of reasons. This allows you to enable "Use date range" so that all data on the Campaign view respects the date range provided. This makes the view faster and more useful for long-running campaigns. This also adds support for query caching (for whoever may use that feature by modifying their config).

**We've been running a variation of this PR (as a patch) in production via mautic-eb since Sept 2018.** It's pretty critical for us, as we have campaigns with ~28 million leads in them. Viewing such a campaign without this is literally impossible.

#### Steps to test this PR:
1. Go to https://mautibox.com/6651/s/config/edit and Campaign Settings.
2. Enable "Use summary statistics" and "Use date range"
3. Progress a contact through any campaign.
4. Go to the campaign https://mautibox.com/6651/s/campaigns and view the Campaign Statistics chart. It should display the events exactly as you would expect.
5. View the Campaign Actions/Conditions/Decisions tabs (if there are any) and they should show the same results they always have as well.
6. Pick a date range with no contacts and hit apply. You'll see the campaign Actions/Conditions/Decisions/Contacts tabs disappear or contain no content. These tabs now respect the date range provided.

### CODE REVIEW

**Comment by heathdutton** [comment]
In my testing just now, I took a page that takes 10~15 *minutes* to load down to ~8s. I'd consider that an improvement. If we merge in (and refactor) #6021 and load contacts by ajax we're talking about a complete transformation.

File: app/bundles/CampaignBundle/Entity/SummaryRepository.php [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                    'FROM_UNIXTIME('.$summary->getDateTriggered()->getTimestamp().')',
+                    $summary->getScheduledCount(),
+                    $summary->getTriggeredCount(),
+                    $summary->getNonActionPathTakenCount(),
+                    $summary->getFailedCount(),
+                ]
+            );
+        }
+
+        $sql = 'INSERT INTO '.MAUTIC_TABLE_PREFIX.'campaign_summary '.
```
Comment (heathdutton) [inline app/bundles/CampaignBundle/Entity/SummaryRepository.php]: Should probably add `DELAYED` to prevent impact of locking.

**Comment by imihandstand** [comment]
Sorry for the question, I'm very new to github: What is the easiest way to implement this code-change? Do I have to manually change all 19 files on my server or is there a simpler way? Thank you!

**Comment by heathdutton** [comment]
@imihandstand if you are using a recent version like 2.14.2+ you may be able to apply this as a git patch.  Go into your mautic folder (not in production) and run `curl -L https://github.com/mautic/mautic/pull/6651.diff | git apply -v`

That will overlay these code changes with your mautic install. 
If you just want to test this code you can go to http://mautibox.com/6551

If you want to be on the safe side, wait till version 2.16.0 which will likely have this in it.

**Comment by heathdutton** [comment]
Looks like I need to do a little refactoring now that pending counts are done differently... :)

**Comment by heathdutton** [comment]
I've resolved conflicts, but note that pending counts are calculated as of 2.15.1 beta (which is good, because it was terrible before), but giving pending figures is impractical when filtering by date, so I've disabled that for now when summary/dates are enabled.

**Comment by heathdutton** [comment]
I've resolved conflicts, but note that pending counts are recalculated as of 2.15.1 beta (which is good). Giving pending figures is impractical when filtering by date, so I've disabled that for now when summary/dates are enabled. May revisit this in the future, but if not, just keep in mind that isn't supported with this optional feature.

We've been running a variation of this PR (as a patch) in production via mautic-eb since Sept 2018. It's pretty critical for us, as we have campaigns with ~28 million leads in them. Viewing such a campaign without this is literally impossible.

**Comment by vicampuzano** [comment]
Hi @heathdutton !!!  :) 
We have 500k+ leads, and also we are experiencing a very slow Muatic behavior. Very very slow... :(

We are in 2.16.0 version but can't see options for summarizing statics.

Email stats table has 23+ GB size, and I think that it could be the reason why our Mautic works slow.

Can you (please, please) help me on how to apply this option in a production environment? 

Thanks!

**Comment by heathdutton** [comment]
> Hi @heathdutton !!! :)
> We have 500k+ leads, and also we are experiencing a very slow Muatic behavior. Very very slow... :(
> 
> We are in 2.16.0 version but can't see options for summarizing statics.
> 
> Email stats table has 23+ GB size, and I think that it could be the reason why our Mautic works slow.
> 
> Can you (please, please) help me on how to apply this option in a production environment?
> 
> Thanks!

You'll want to `curl -L https://github.com/mautic/mautic/pull/6651.diff | git apply -v` inside your mautic folder, then run the `php console mautic:campaign:summarize` command to backfill data. Then go into Settings > Campaign Settings and enable "Use summary statistics" and "Use date range".

I highly recommend you test this locally or on a dev server before attempting this in production.

**Comment by heathdutton** [comment]
> Hi @heathdutton !!! :)
> We have 500k+ leads, and also we are experiencing a very slow Muatic behavior. Very very slow... :(
> 
> We are in 2.16.0 version but can't see options for summarizing statics.
> 
> Email stats table has 23+ GB size, and I think that it could be the reason why our Mautic works slow.
> 
> Can you (please, please) help me on how to apply this option in a production environment?
> 
> Thanks!

You'll want to `curl -L https://github.com/mautic/mautic/pull/6651.diff | git apply -v` inside your mautic folder, then run the `php console mautic:campaign:summarize` command to backfill data. Then go into Settings > Campaign Settings and enable "Use summary statistics" and "Use date range".

I highly recommend you test this locally or on a dev server before attempting this in production.

**Comment by vicampuzano** [comment]
> > Hi @heathdutton !!! :)
> > We have 500k+ leads, and also we are experiencing a very slow Muatic behavior. Very very slow... :(
> > We are in 2.16.0 version but can't see options for summarizing statics.
> > Email stats table has 23+ GB size, and I think that it could be the reason why our Mautic works slow.
> > Can you (please, please) help me on how to apply this option in a production environment?
> > Thanks!
> 
> You'll want to `curl -L https://github.com/mautic/mautic/pull/6651.diff | git apply -v` inside your mautic folder, then run the `php console mautic:campaign:summarize` command to backfill data. Then go into Settings > Campaign Settings and enable "Use summary statistics" and "Use date range".
> 
> I highly recommend you test this locally or on a dev server before attempting this in production.

Thanks @heathdutton !!! ;) 

Your help is very useful... Thanks thanks!

I applied the patch but it doesn't work in version 2.16.0. Seems web inteerface is working but at console every command returns this error:

```
[Symfony\Component\Debug\Exception\ContextErrorException]                                                            
  Catchable Fatal Error: Argument 5 passed to Mautic\CampaignBundle\Executioner\Logger\EventLogger::__construct() mus  
  t be an instance of Mautic\CampaignBundle\Model\SummaryModel, none given, called in /home/admin/web/mt.growthup.es/  
  public_html/app/cache/prod/appProdProjectContainer.php on line 5844 and defined
```

I'm not sure how to continue... Can you help us?

Thanks again!

**Comment by vicampuzano** [comment]
Ok. sorry. 

This error is caused because I executed commands as root user. I manually cleared cache and fixed permissions and it worked. 

But for now I'm trying to solve another issue:

```
[Doctrine\DBAL\Exception\TableNotFoundException]                                                                     
  An exception occurred while executing 'SELECT cs.date_triggered FROM campaign_summary cs ORDER BY cs.date_triggered  
   ASC LIMIT 1':                                                                                                       
                                                                                                                       
  SQLSTATE[42S02]: Base table or view not found: 1146 Table 'admin_mautic.campaign_summary' doesn't exist 
```

Seems the table is not created but there isn't migration to apply...

**Comment by heathdutton** [comment]
> Seems the table is not created but there isn't migration to apply...

Oh wow, can't believe I forgot to drop in a migration 😢   
For now:
```
CREATE TABLE `campaign_summary` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `campaign_id` int(11) DEFAULT NULL,
  `event_id` int(11) NOT NULL,
  `date_triggered` datetime DEFAULT NULL COMMENT '(DC2Type:datetime)',
  `scheduled_count` int(11) NOT NULL,
  `triggered_count` int(11) NOT NULL,
  `non_action_path_taken_count` int(11) NOT NULL,
  `failed_count` int(11) NOT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `campaign_event_date_triggered` (`campaign_id`,`event_id`,`date_triggered`),
  KEY `IDX_6692FA4FF639F774` (`campaign_id`),
  KEY `IDX_6692FA4F71F7E88B` (`event_id`),
  CONSTRAINT `FK_6692FA4F71F7E88B` FOREIGN KEY (`event_id`) REFERENCES `campaign_events` (`id`) ON DELETE CASCADE,
  CONSTRAINT `FK_6692FA4FF639F774` FOREIGN KEY (`campaign_id`) REFERENCES `campaigns` (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=0 DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
```

**Comment by npracht** [comment]
Hi there!
It has been decided to not create any extra Mautic 2 minor versions (which means no more features in M2) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.1.0 candidate.**

## How to do?
1. Check if your feature or enhancement is still missing and relevant in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please consider closing your PR if you are absolutely sure that the feature has made it into Mautic 3 already

**Please report results by commenting on your PR to make us administration easier.**
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by RCheesley** [comment]
@heathdutton could you add the migration into this PR and address the conflicts? Seems it could be a useful one for large instances so we could look to include this in 3.2 which would mean we'd need this to be done by **Sunday, 15th November** at the latest.

**Comment by escopecz** [comment]
@heathdutton this looks great! I hope you don't mind if I rebase and add the migration.

**Comment by escopecz** [comment]
Oops, it conflicts with https://github.com/mautic/mautic/pull/6907 which will probably need more changes added to this PR. I'll finish the rebase and see what happens.

**Comment by escopecz** [comment]
@heathdutton is @scottshipman still in your team? Could you ask him to sign CLA?

File: app/bundles/CampaignBundle/Entity/SummaryRepository.php [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            '        FROM_UNIXTIME(UNIX_TIMESTAMP(t.date_triggered) - (UNIX_TIMESTAMP(t.date_triggered) % 3600)) AS date_triggered, '.
+            '        SUM(IF(t.is_scheduled = 1 AND t.trigger_date > NOW(), 1, 0)) as scheduled_count, '.
+            '        SUM(IF(t.is_scheduled = 1 AND t.trigger_date > NOW(), 0, t.non_action_path_taken)) as non_action_path_taken_count, '.
+            '        SUM(IF((t.is_scheduled = 1 AND t.trigger_date > NOW()) OR t.non_action_path_taken, 0, fe.log_id IS NOT NULL)) as failed_count, '.
+            '        SUM(IF((t.is_scheduled = 1 AND t.trigger_date > NOW()) OR t.non_action_path_taken OR fe.log_id IS NOT NULL, 0, 1)) as triggered_count '.
+            '    FROM '.MAUTIC_TABLE_PREFIX.'campaign_lead_event_log t '.
+            '    LEFT JOIN '.MAUTIC_TABLE_PREFIX.'campaign_lead_event_failed_log fe '.
+            '        ON fe.log_id = t.id '.
+            '    WHERE (t.date_triggered BETWEEN FROM_UNIXTIME(:dateFrom) AND FROM_UNIXTIME(:dateTo)) '.
+            '    GROUP BY campaign_id, event_id, date_triggered '.
```
Comment (escopecz) [inline app/bundles/CampaignBundle/Entity/SummaryRepository.php:144]: @heathdutton I wrote a basic functional test to cover the SummarizeCommand but it does not work as I'd expect. It fails on https://github.com/mautic/mautic/pull/6651/files#diff-c73ead79e78e7652c06c382e160af659c4e6eccca3341bd21834c9c127988768R74

I've added to lead event logs entries for the same campaign, same event, same hour but 2 different contacts. It counts those 2 contacts as 1. It should be 2, right? The summary rows are generated by this query which I haven't touched.

**Comment by escopecz** [comment]
Closing this PR for it's younger and fixed version at https://github.com/mautic/mautic/pull/9471
